consensus/src/terminator.rs

```diff
@@ -0,0 +1,121 @@
+use futures::channel::oneshot;
+use log::debug;
+
+pub type TerminatorConnection = (oneshot::Sender<()>, oneshot::Receiver<()>);
+pub type ShutdownConnection = (oneshot::Receiver<()>, Option<TerminatorConnection>);
+
+/// Struct that holds connections to offspring and parent components/tasks
+/// and enables a clean/synchronized shutdown
+pub struct Terminator {
+    component_name: String,
+    parent_connection: Option<TerminatorConnection>,
+    offspring_connections: Vec<TerminatorConnection>,
+    offspring_names: Vec<String>,
+}
+
+impl Terminator {
+    pub fn new(parent_connection: Option<TerminatorConnection>, component_name: &str) -> Self {
+        Self {
+            component_name: String::from(component_name),
+            parent_connection,
+            offspring_connections: Vec::<_>::new(),
+            offspring_names: Vec::<_>::new(),
+        }
+    }
+
+    /// Add a connection to an offspring component/task
+    pub fn add_offspring_connection(&mut self, name: &str) -> TerminatorConnection {
+        let (sender, offspring_recv) = oneshot::channel();
+        let (offspring_sender, recv) = oneshot::channel();
+
+        let endpoint = (sender, recv);
+        let offspring_endpoint = (offspring_sender, offspring_recv);
+
+        self.offspring_connections.push(endpoint);
+        self.offspring_names.push(name.to_string());
+        offspring_endpoint
+    }
+
+    /// Perform a synchronized shutdown
+    pub async fn terminate_sync(self) {
+        let (offspring_senders, offspring_receivers): (Vec<_>, Vec<_>) =
+            self.offspring_connections.into_iter().unzip();
+        let offspring_senders: Vec<_> = offspring_senders
+            .into_iter()
+            .zip(self.offspring_names.clone().into_iter())
+            .collect();
+        let offspring_receivers: Vec<_> = offspring_receivers
+            .into_iter()
+            .zip(self.offspring_names.clone().into_iter())
+            .collect();
+
+        debug!(
+            target: &self.component_name,
+            "Terminator preparing for shutdown.",
+        );
+
+        // Make sure that all descendants recieved exit and won't be communicating with other components
+        for (receiver, name) in offspring_receivers {
+            if receiver.await.is_err() {
+                debug!(
+                    target: &self.component_name,
+                    "Terminator failed to receive from {}. Clean shutdown unsuccessful.",
+                    name,
+                );
+                return;
```

Comment:
Reconsider all of those `early exits` - perhaps it would be better to log all failed services. Probably requires some verification of whether it won't block forever.

Reply:
Pretty sure it won't block - got rid of them.

Reply:
Following our recent conversation with @fixxxedpoint: there are cases where it ends with early exits and blocks without them - it can happen when some component calls terminate_sync after it's child task exits with an error (which can block the system forever - at least in principle).